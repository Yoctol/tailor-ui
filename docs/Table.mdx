---
name: Table
menu: Components
route: /components/table
---

import { PropsTable, Playground } from 'docz';
import { range } from 'ramda';
import { Box, Button, Flex, Heading, Table, Space } from 'tailor-ui';

# Table

A table displays rows of data.

## When To Use

- To display a collection of structured data.
- To sort, search, paginate, filter data.

## Examples

```js
import { Table } from 'tailor-ui';
```

### Basic Table

<Playground>
  <Box width="100%" height="100%" borderRadius="xl" bg="#eef0f5" p="6">
    <Table>
      <Table.Head>
        <Table.HeadColumn>Name</Table.HeadColumn>
        <Table.HeadColumn>Platform</Table.HeadColumn>
        <Table.HeadColumn>People</Table.HeadColumn>
        <Table.HeadColumn>Ammount</Table.HeadColumn>
        <Table.HeadColumn>Action</Table.HeadColumn>
      </Table.Head>
      <Table.Body>
        {range(1, 4).map(value => (
          <Table.Row key={value}>
            <Table.Column>Name {value}</Table.Column>
            <Table.Column>Platform {value}</Table.Column>
            <Table.Column>10,000</Table.Column>
            <Table.Column>1,000 NTD</Table.Column>
            <Table.Column>
              <Button variant="primary-invert">Action</Button>
            </Table.Column>
          </Table.Row>
        ))}
      </Table.Body>
    </Table>
  </Box>
</Playground>

### Header & Footer

<Playground>
  <Box width="100%" height="100%" borderRadius="xl" bg="#eef0f5" p="6">
    <Table
      header={
        <Space py="24px" px="6">
          <Heading.h4>Text Here</Heading.h4>
        </Space>
      }
      footer={
        <Space py="24px" px="6">
          <Button variant="regular">+ Button</Button>
        </Space>
      }
    >
      <Table.Head>
        <Table.HeadColumn>Name</Table.HeadColumn>
        <Table.HeadColumn>Platform</Table.HeadColumn>
        <Table.HeadColumn>People</Table.HeadColumn>
        <Table.HeadColumn>Ammount</Table.HeadColumn>
        <Table.HeadColumn>Action</Table.HeadColumn>
      </Table.Head>
      <Table.Body>
        {range(1, 4).map(value => (
          <Table.Row key={value}>
            <Table.Column>Name {value}</Table.Column>
            <Table.Column>Platform {value}</Table.Column>
            <Table.Column>10,000</Table.Column>
            <Table.Column>1,000 NTD</Table.Column>
            <Table.Column>
              <Button variant="primary-invert">Action</Button>
            </Table.Column>
          </Table.Row>
        ))}
      </Table.Body>
    </Table>
  </Box>
</Playground>

### Complex Footer

<Playground>
  <Box width="100%" height="100%" borderRadius="xl" bg="#eef0f5" p="6">
    <Table
      footer={
        <div>
          <Flex
            justifyContent="flex-end"
            alignItems="center"
            height="80px"
            px="6"
          >
            <Space>
              <Heading.h3>Text Here</Heading.h3>
              <Heading.h3>Text Here</Heading.h3>
            </Space>
          </Flex>
          <Flex
            bg="primary"
            justifyContent="flex-end"
            alignItems="center"
            height="80px"
            px="6"
          >
            <Button variant="primary-invert" rounded>
              Action
            </Button>
          </Flex>
        </div>
      }
    >
      <Table.Head>
        <Table.HeadColumn>Name</Table.HeadColumn>
        <Table.HeadColumn>Platform</Table.HeadColumn>
        <Table.HeadColumn>People</Table.HeadColumn>
        <Table.HeadColumn>Ammount</Table.HeadColumn>
        <Table.HeadColumn>Action</Table.HeadColumn>
      </Table.Head>
      <Table.Body>
        {range(1, 4).map(value => (
          <Table.Row key={value}>
            <Table.Column>Name {value}</Table.Column>
            <Table.Column>Platform {value}</Table.Column>
            <Table.Column>10,000</Table.Column>
            <Table.Column>1,000 NTD</Table.Column>
            <Table.Column>
              <Button variant="primary-invert">Action</Button>
            </Table.Column>
          </Table.Row>
        ))}
      </Table.Body>
    </Table>
  </Box>
</Playground>

## API

<PropsTable of={Table} />
