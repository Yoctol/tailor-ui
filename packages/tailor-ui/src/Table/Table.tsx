import React, { FunctionComponent, PureComponent } from 'react';
import styled from 'styled-components';
import {
  BorderColorProps,
  BordersProps,
  ColorProps,
  SpaceProps,
  TextAlignProps,
  WidthProps,
  borderColor,
  borders,
  color,
  space,
  textAlign,
  width,
} from 'styled-system';

import tag from '../utils/CleanTag';

type HeadColumnProps = WidthProps & SpaceProps & BordersProps & ColorProps;

const HeadColumn = styled(tag.th)<HeadColumnProps>`
  padding: ${p => p.theme.paddings.xs} ${p => p.theme.paddings.sm};
  border-color: ${p => p.theme.colors.gray300};
  font-weight: 500;

  ${width};
  ${space};
  ${borders};
  ${color};
`;

type ColumnProps = SpaceProps & BordersProps & BorderColorProps & ColorProps;

const Column = styled(tag.td)<ColumnProps>`
  padding: ${p => p.theme.paddings.xs} ${p => p.theme.paddings.sm};

  ${space};
  ${borders};
  ${borderColor};
  ${color};
`;

Column.defaultProps = {
  borderColor: 'gray300',
};

type RowProps = ColorProps;

const Row = styled(tag.tr)<RowProps>`
  border-bottom: ${p => p.theme.borders.base} ${p => p.theme.colors.gray300};

  ${color};
`;

const Head: FunctionComponent = ({ children }) => (
  <thead>
    <Row>{children}</Row>
  </thead>
);

const Body: FunctionComponent = ({ children }) => <tbody>{children}</tbody>;

type StyledTableProps = TextAlignProps & WidthProps;

const StyledTable = styled(tag.table)<StyledTableProps>`
  overflow: hidden;
  border-spacing: 0;
  border-collapse: collapse;
  border-style: hidden;
  border-top-left-radius: ${p => p.theme.radii.base};
  border-top-right-radius: ${p => p.theme.radii.base};
  box-shadow: 0 0 0 1px ${p => p.theme.colors.gray300};
  font-size: ${p => p.theme.fontSizes.sm};

  & > thead {
    border-bottom: ${p => p.theme.borders.base} ${p => p.theme.colors.gray300};
    background-color: ${p => p.theme.colors.gray100};
  }

  ${width};
  ${textAlign};
`;

class Table extends PureComponent<StyledTableProps> {
  static Head = Head;

  static HeadColumn = HeadColumn;

  static Body = Body;

  static Row = Row;

  static Column = Column;

  static defaultProps = {
    width: '100%',
    textAlign: 'left',
  };

  render() {
    return <StyledTable {...this.props} />;
  }
}

export default Table;
