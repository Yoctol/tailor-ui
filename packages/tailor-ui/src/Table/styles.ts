import styled, { StyledComponent, css } from 'styled-components';
import { TextAlignProps, WidthProps, textAlign, width } from 'styled-system';

export const HeadColumn: StyledComponent<'th', any, WidthProps> = styled.th`
  padding: ${p => p.theme.paddings.xs} ${p => p.theme.paddings.sm};
  font-weight: bold;

  ${width};
`;

export const Column = styled.td`
  padding: ${p => p.theme.paddings.xs} ${p => p.theme.paddings.sm};
`;

export const Row = styled.tr`
  height: 56px;
  border-bottom: ${p => p.theme.borders.base} ${p => p.theme.colors.gray300};

  ${p => p.theme.transition};
`;

export type StyledTableProps = TextAlignProps & WidthProps;

export const StyledTable = styled.table<StyledTableProps>`
  flex: none;
  overflow: hidden;
  border-spacing: 0;
  border-collapse: collapse;
  border-style: hidden;
  border-radius: ${p => p.theme.radii.xl};
  font-size: ${p => p.theme.fontSizes.base};

  & > thead {
    border-bottom: ${p => p.theme.borders.base} ${p => p.theme.colors.gray300};
    background-color: ${p => p.theme.colors.surface2};
  }

  & > tbody > ${Row /* sc-selector */} {
    background-color: ${p => p.theme.colors.light};

    &:hover {
      background-color: ${p => p.theme.colors.gray100};
    }
  }

  ${width};
  ${textAlign};
`;

export type TableWrapperProps = WidthProps & {
  hasHeader: boolean;
  hasFooter: boolean;
};

export const TableWrapper = styled.div<TableWrapperProps>`
  flex: none;
  overflow: hidden;
  border-radius: ${p => p.theme.radii.xl};
  background-color: ${p => p.theme.colors.light};
  box-shadow: ${p => p.theme.shadows.base};

  /* stylelint-disable-next-line no-descending-specificity */
  & > ${StyledTable} {
    width: 100%;
    border-radius: 0;

    ${p =>
      p.hasHeader &&
      css`
        border-top: ${p.theme.borders.base} ${p.theme.colors.gray300};
      `}

    ${p =>
      p.hasFooter &&
      css`
        border-bottom: ${p.theme.borders.base} ${p.theme.colors.gray300};
      `}
  }

  ${width};
`;
