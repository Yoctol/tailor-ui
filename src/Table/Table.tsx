import React, { ComponentPropsWithoutRef, ReactNode, forwardRef } from 'react';

import {
  Column,
  HeadColumn,
  Row,
  StyledTable,
  StyledTableProps,
  TableWrapper,
} from './styles';

const Head = forwardRef<HTMLTableSectionElement, { children: ReactNode }>(
  function Head({ children }, ref) {
    return (
      <thead ref={ref}>
        <Row>{children}</Row>
      </thead>
    );
  }
);

const Body = forwardRef<HTMLTableSectionElement, { children: ReactNode }>(
  function Body({ children }, ref) {
    return <tbody ref={ref}>{children}</tbody>;
  }
);

export type TableProps = ComponentPropsWithoutRef<'table'> &
  Omit<StyledTableProps, 'hasHeader' | 'hasFooter'> & {
    header?: ReactNode;
    footer?: ReactNode;
  };

const Table = forwardRef<HTMLTableElement, TableProps>(function Table(
  { header, footer, width = '100%', textAlign = 'center', ...props },
  ref
) {
  const optionsProps = {
    hasHeader: Boolean(header),
    hasFooter: Boolean(footer),
  };
  const table = (
    <StyledTable
      ref={ref}
      width={width as string}
      textAlign={textAlign}
      {...optionsProps}
      {...props}
    />
  );

  if (header || footer) {
    return (
      <TableWrapper ref={ref} width={width} {...optionsProps} {...props}>
        {header}
        {table}
        {footer}
      </TableWrapper>
    );
  }

  return table;
}) as React.ForwardRefExoticComponent<
  TableProps & React.RefAttributes<HTMLTableElement>
> & {
  Head: typeof Head;
  HeadColumn: typeof HeadColumn;
  Body: typeof Body;
  Row: typeof Row;
  Column: typeof Column;
};

Table.Head = Head;
Table.HeadColumn = HeadColumn;
Table.Body = Body;
Table.Row = Row;
Table.Column = Column;

export { Table };
