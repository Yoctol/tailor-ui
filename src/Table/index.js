import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import {
  themeGet,
  width,
  space,
  textAlign,
  borders,
  borderColor,
} from 'styled-system';

import theme from '../theme';

const HeadColumn = styled.th`
  padding: ${themeGet('space.paddingYLg')} ${themeGet('space.paddingXLg')};
  font-weight: 500;

  ${width};
  ${space};
  ${borders};
  ${borderColor};
`;

HeadColumn.propTypes = {
  ...width.propTypes,
  ...space.propTypes,
  ...borders.propTypes,
  ...borderColor.propTypes,
};

HeadColumn.defaultProps = {
  theme,
  borderColor: 'border',
};

const Column = styled.td`
  padding: ${themeGet('space.paddingYLg')} ${themeGet('space.paddingXLg')};

  ${borders};
  ${borderColor};
`;

Column.propTypes = {
  ...space.propTypes,
  ...borders.propTypes,
  ...borderColor.propTypes,
};

Column.defaultProps = {
  theme,
  borderColor: 'border',
};

const Row = styled.tr`
  border-bottom: ${themeGet('borders.default')} ${themeGet('colors.border')};
`;

Row.defaultProps = {
  theme,
};

const Head = ({ children }) => (
  <thead>
    <Row>{children}</Row>
  </thead>
);

Head.propTypes = {
  children: PropTypes.element.isRequired,
};

const Body = ({ children }) => <tbody>{children}</tbody>;

Body.propTypes = {
  children: PropTypes.element,
};

Body.defaultProps = {
  children: '',
};

const Table = styled.table`
  overflow: hidden;
  border-spacing: 0;
  border-collapse: collapse;
  border-style: hidden;
  border-top-left-radius: ${themeGet('radii.1')};
  border-top-right-radius: ${themeGet('radii.1')};
  box-shadow: 0 0 0 1px ${themeGet('colors.border')};

  & > thead {
    border-bottom: ${themeGet('borders.default')} ${themeGet('colors.border')};
    background-color: ${themeGet('colors.gray.9')};
  }

  ${textAlign};
  ${width};
`;

Table.defaultProps = {
  theme,
  width: '100%',
  textAlign: 'left',
};

Table.Head = Head;
Table.HeadColumn = HeadColumn;
Table.Body = Body;
Table.Row = Row;
Table.Column = Column;

export default Table;
