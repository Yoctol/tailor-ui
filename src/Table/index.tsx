import React, { PureComponent, SFC } from 'react';
import styled from 'styled-components';
import {
  BorderColorProps,
  BordersProps,
  SpaceProps,
  TextAlignProps,
  WidthProps,
  borderColor,
  borders,
  space,
  textAlign,
  themeGet,
  width,
} from 'styled-system';

type HeadColumnProps = WidthProps &
  SpaceProps &
  BordersProps &
  BorderColorProps;

const HeadColumn = styled<HeadColumnProps, 'th'>('th')`
  padding: ${themeGet('paddings.xs')} ${themeGet('paddings.sm')};
  font-weight: 500;

  ${width};
  ${space};
  ${borders};
  ${borderColor};
`;

HeadColumn.defaultProps = {
  borderColor: 'gray.8',
};

type ColumnProps = SpaceProps & BordersProps & BorderColorProps;

const Column = styled<ColumnProps, 'td'>('td').attrs({
  borderColor: 'gray.8',
})`
  padding: ${themeGet('paddings.xs')} ${themeGet('paddings.sm')};

  ${space} ${borders};
  ${borderColor};
`;

const Row = styled.tr`
  border-bottom: ${themeGet('borders.base')} ${themeGet('colors.gray.8')};
`;

const Head: SFC = ({ children }) => (
  <thead>
    <Row>{children}</Row>
  </thead>
);

const Body: SFC = ({ children }) => <tbody>{children}</tbody>;

type StyledTableProps = TextAlignProps & WidthProps;

const StyledTable = styled<StyledTableProps, 'table'>('table')`
  overflow: hidden;
  border-spacing: 0;
  border-collapse: collapse;
  border-style: hidden;
  border-top-left-radius: ${themeGet('radii.base')};
  border-top-right-radius: ${themeGet('radii.base')};
  box-shadow: 0 0 0 1px ${themeGet('colors.gray.8')};
  font-size: ${themeGet('fontSizes.sm')};

  & > thead {
    border-bottom: ${themeGet('borders.base')} ${themeGet('colors.gray.8')};
    background-color: ${themeGet('colors.gray.9')};
  }

  ${width};
  ${textAlign};
`;

class Table extends PureComponent<StyledTableProps> {
  static Head = Head;

  static HeadColumn = HeadColumn;

  static Body = Body;

  static Row = Row;

  static Column = Column;

  static defaultProps = {
    width: '100%',
    textAlign: 'left',
  };

  render() {
    return <StyledTable {...this.props} />;
  }
}

export default Table;
