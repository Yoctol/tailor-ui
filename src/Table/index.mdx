---
name: Table
menu: Components
route: /components/table
---

import { PropsTable, Playground } from 'docz';
import { range } from 'ramda';
import Table from './';

# Table

A table displays rows of data.

## When To Use
- To display a collection of structured data.
- To sort, search, paginate, filter data.

## Examples

```js
import { Table } from '@yoctol/ui';
```

### Basic Table

<Playground>
  <Table>
    <Table.Head>
      {range(2, 10).map(x => <Table.HeadColumn key={x}>{x}</Table.HeadColumn>)}
    </Table.Head>
    <Table.Body>
      {range(1, 10).map(x => (
        <Table.Row key={x}>
          {range(2, 10).map(y => (
            <Table.Column key={y}>
              {y} * {x} = {x * y}
            </Table.Column>
          ))}
        </Table.Row>
      ))}
    </Table.Body>
  </Table>
</Playground>

### Custom Table

<Playground>
  <Table textAlign="center">
    <Table.Head>
      <Table.HeadColumn colSpan="8">Time</Table.HeadColumn>
    </Table.Head>
    <Table.Body>
      <Table.Row>
        <Table.Column />
        <Table.Column>日</Table.Column>
        <Table.Column>一</Table.Column>
        <Table.Column>二</Table.Column>
        <Table.Column>三</Table.Column>
        <Table.Column>四</Table.Column>
        <Table.Column>五</Table.Column>
        <Table.Column>六</Table.Column>
      </Table.Row>
      <Table.Row>
        <Table.Column>
          Morning<br />10:00-13:00
        </Table.Column>
        <Table.Column
          rowSpan="3"
          borderColor="gray300"
          borderLeft="base"
          borderRight="base"
        >
          Day off
        </Table.Column>
        {range(1, 7).map(key => (
          <Table.Column key={key}>
            <div>Master {key}</div>
            <div>Slave {key}</div>
          </Table.Column>
        ))}
      </Table.Row>
      <Table.Row>
        <Table.Column>
          Afternoon<br />14:00-18:00
        </Table.Column>
        {range(1, 7).map(key => (
          <Table.Column key={key}>
            <div>Master {key}</div>
            <div>Slave {key}</div>
          </Table.Column>
        ))}
      </Table.Row>
      <Table.Row>
        <Table.Column>
          Night<br />19:00-22:00
        </Table.Column>
        {range(1, 7).map(key => (
          <Table.Column key={key}>
            <div>Master {key}</div>
            <div>Slave {key}</div>
          </Table.Column>
        ))}
      </Table.Row>
    </Table.Body>
  </Table>
</Playground>

## API

<PropsTable of={Table} />
