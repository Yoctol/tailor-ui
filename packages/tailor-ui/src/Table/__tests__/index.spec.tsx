import React from 'react';

import { render } from 'test/test-utils';

import Table from '../Table';

function range(start: number, end: number) {
  return new Array(end - start).fill(0).map((_, i) => start + i);
}

describe('Table', () => {
  it('should render table correctly', () => {
    const { container } = render(
      <Table>
        <Table.Head>
          <Table.HeadColumn>Name</Table.HeadColumn>
          <Table.HeadColumn>Platform</Table.HeadColumn>
          <Table.HeadColumn>People</Table.HeadColumn>
          <Table.HeadColumn>Ammount</Table.HeadColumn>
          <Table.HeadColumn>Action</Table.HeadColumn>
        </Table.Head>
        <Table.Body>
          {range(1, 4).map(value => (
            <Table.Row key={value}>
              <Table.Column>Name {value}</Table.Column>
              <Table.Column>Platform {value}</Table.Column>
              <Table.Column>10,000</Table.Column>
              <Table.Column>1,000 NTD</Table.Column>
              <Table.Column>Action</Table.Column>
            </Table.Row>
          ))}
        </Table.Body>
      </Table>
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render table correctly with header and footer', () => {
    const { container } = render(
      <Table header="Header" footer="Footer">
        <Table.Head>
          <Table.HeadColumn>Name</Table.HeadColumn>
          <Table.HeadColumn>Platform</Table.HeadColumn>
          <Table.HeadColumn>People</Table.HeadColumn>
          <Table.HeadColumn>Ammount</Table.HeadColumn>
          <Table.HeadColumn>Action</Table.HeadColumn>
        </Table.Head>
        <Table.Body>
          {range(1, 4).map(value => (
            <Table.Row key={value}>
              <Table.Column>Name {value}</Table.Column>
              <Table.Column>Platform {value}</Table.Column>
              <Table.Column>10,000</Table.Column>
              <Table.Column>1,000 NTD</Table.Column>
              <Table.Column>Action</Table.Column>
            </Table.Row>
          ))}
        </Table.Body>
      </Table>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
