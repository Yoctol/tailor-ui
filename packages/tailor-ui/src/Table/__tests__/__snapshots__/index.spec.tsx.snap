// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render multiplication table correctly 1`] = `
.c2 {
  padding: 0.5rem 0.75rem;
  border-color: #e0e0e0;
  font-weight: 500;
}

.c3 {
  padding: 0.5rem 0.75rem;
  border-color: #e0e0e0;
  border-color: #e0e0e0;
}

.c1 {
  border-bottom: 1px solid #e0e0e0;
}

.c0 {
  overflow: hidden;
  border-spacing: 0;
  border-collapse: collapse;
  border-style: hidden;
  border-top-left-radius: 0.25rem;
  border-top-right-radius: 0.25rem;
  box-shadow: 0 0 0 1px #e0e0e0;
  font-size: 0.875rem;
  width: 100%;
  text-align: left;
}

.c0 > thead {
  border-bottom: 1px solid #e0e0e0;
  background-color: #fafafa;
}

<table
  class="c0"
  width="100%"
>
  <thead>
    <tr
      class="c1"
    >
      <th
        class="c2"
      >
        2
      </th>
      <th
        class="c2"
      >
        3
      </th>
      <th
        class="c2"
      >
        4
      </th>
      <th
        class="c2"
      >
        5
      </th>
      <th
        class="c2"
      >
        6
      </th>
      <th
        class="c2"
      >
        7
      </th>
      <th
        class="c2"
      >
        8
      </th>
      <th
        class="c2"
      >
        9
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        1
         = 
        2
      </td>
      <td
        class="c3"
      >
        3
         * 
        1
         = 
        3
      </td>
      <td
        class="c3"
      >
        4
         * 
        1
         = 
        4
      </td>
      <td
        class="c3"
      >
        5
         * 
        1
         = 
        5
      </td>
      <td
        class="c3"
      >
        6
         * 
        1
         = 
        6
      </td>
      <td
        class="c3"
      >
        7
         * 
        1
         = 
        7
      </td>
      <td
        class="c3"
      >
        8
         * 
        1
         = 
        8
      </td>
      <td
        class="c3"
      >
        9
         * 
        1
         = 
        9
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        2
         = 
        4
      </td>
      <td
        class="c3"
      >
        3
         * 
        2
         = 
        6
      </td>
      <td
        class="c3"
      >
        4
         * 
        2
         = 
        8
      </td>
      <td
        class="c3"
      >
        5
         * 
        2
         = 
        10
      </td>
      <td
        class="c3"
      >
        6
         * 
        2
         = 
        12
      </td>
      <td
        class="c3"
      >
        7
         * 
        2
         = 
        14
      </td>
      <td
        class="c3"
      >
        8
         * 
        2
         = 
        16
      </td>
      <td
        class="c3"
      >
        9
         * 
        2
         = 
        18
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        3
         = 
        6
      </td>
      <td
        class="c3"
      >
        3
         * 
        3
         = 
        9
      </td>
      <td
        class="c3"
      >
        4
         * 
        3
         = 
        12
      </td>
      <td
        class="c3"
      >
        5
         * 
        3
         = 
        15
      </td>
      <td
        class="c3"
      >
        6
         * 
        3
         = 
        18
      </td>
      <td
        class="c3"
      >
        7
         * 
        3
         = 
        21
      </td>
      <td
        class="c3"
      >
        8
         * 
        3
         = 
        24
      </td>
      <td
        class="c3"
      >
        9
         * 
        3
         = 
        27
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        4
         = 
        8
      </td>
      <td
        class="c3"
      >
        3
         * 
        4
         = 
        12
      </td>
      <td
        class="c3"
      >
        4
         * 
        4
         = 
        16
      </td>
      <td
        class="c3"
      >
        5
         * 
        4
         = 
        20
      </td>
      <td
        class="c3"
      >
        6
         * 
        4
         = 
        24
      </td>
      <td
        class="c3"
      >
        7
         * 
        4
         = 
        28
      </td>
      <td
        class="c3"
      >
        8
         * 
        4
         = 
        32
      </td>
      <td
        class="c3"
      >
        9
         * 
        4
         = 
        36
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        5
         = 
        10
      </td>
      <td
        class="c3"
      >
        3
         * 
        5
         = 
        15
      </td>
      <td
        class="c3"
      >
        4
         * 
        5
         = 
        20
      </td>
      <td
        class="c3"
      >
        5
         * 
        5
         = 
        25
      </td>
      <td
        class="c3"
      >
        6
         * 
        5
         = 
        30
      </td>
      <td
        class="c3"
      >
        7
         * 
        5
         = 
        35
      </td>
      <td
        class="c3"
      >
        8
         * 
        5
         = 
        40
      </td>
      <td
        class="c3"
      >
        9
         * 
        5
         = 
        45
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        6
         = 
        12
      </td>
      <td
        class="c3"
      >
        3
         * 
        6
         = 
        18
      </td>
      <td
        class="c3"
      >
        4
         * 
        6
         = 
        24
      </td>
      <td
        class="c3"
      >
        5
         * 
        6
         = 
        30
      </td>
      <td
        class="c3"
      >
        6
         * 
        6
         = 
        36
      </td>
      <td
        class="c3"
      >
        7
         * 
        6
         = 
        42
      </td>
      <td
        class="c3"
      >
        8
         * 
        6
         = 
        48
      </td>
      <td
        class="c3"
      >
        9
         * 
        6
         = 
        54
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        7
         = 
        14
      </td>
      <td
        class="c3"
      >
        3
         * 
        7
         = 
        21
      </td>
      <td
        class="c3"
      >
        4
         * 
        7
         = 
        28
      </td>
      <td
        class="c3"
      >
        5
         * 
        7
         = 
        35
      </td>
      <td
        class="c3"
      >
        6
         * 
        7
         = 
        42
      </td>
      <td
        class="c3"
      >
        7
         * 
        7
         = 
        49
      </td>
      <td
        class="c3"
      >
        8
         * 
        7
         = 
        56
      </td>
      <td
        class="c3"
      >
        9
         * 
        7
         = 
        63
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        8
         = 
        16
      </td>
      <td
        class="c3"
      >
        3
         * 
        8
         = 
        24
      </td>
      <td
        class="c3"
      >
        4
         * 
        8
         = 
        32
      </td>
      <td
        class="c3"
      >
        5
         * 
        8
         = 
        40
      </td>
      <td
        class="c3"
      >
        6
         * 
        8
         = 
        48
      </td>
      <td
        class="c3"
      >
        7
         * 
        8
         = 
        56
      </td>
      <td
        class="c3"
      >
        8
         * 
        8
         = 
        64
      </td>
      <td
        class="c3"
      >
        9
         * 
        8
         = 
        72
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        2
         * 
        9
         = 
        18
      </td>
      <td
        class="c3"
      >
        3
         * 
        9
         = 
        27
      </td>
      <td
        class="c3"
      >
        4
         * 
        9
         = 
        36
      </td>
      <td
        class="c3"
      >
        5
         * 
        9
         = 
        45
      </td>
      <td
        class="c3"
      >
        6
         * 
        9
         = 
        54
      </td>
      <td
        class="c3"
      >
        7
         * 
        9
         = 
        63
      </td>
      <td
        class="c3"
      >
        8
         * 
        9
         = 
        72
      </td>
      <td
        class="c3"
      >
        9
         * 
        9
         = 
        81
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table should render week schedule 1`] = `
.c2 {
  padding: 0.5rem 0.75rem;
  border-color: #e0e0e0;
  font-weight: 500;
}

.c3 {
  padding: 0.5rem 0.75rem;
  border-color: #e0e0e0;
  border-color: #e0e0e0;
}

.c4 {
  padding: 0.5rem 0.75rem;
  border-color: #e0e0e0;
  border-right: 1px solid;
  border-left: 1px solid;
  border-color: #e0e0e0;
}

.c1 {
  border-bottom: 1px solid #e0e0e0;
}

.c0 {
  overflow: hidden;
  border-spacing: 0;
  border-collapse: collapse;
  border-style: hidden;
  border-top-left-radius: 0.25rem;
  border-top-right-radius: 0.25rem;
  box-shadow: 0 0 0 1px #e0e0e0;
  font-size: 0.875rem;
  width: 100%;
  text-align: center;
}

.c0 > thead {
  border-bottom: 1px solid #e0e0e0;
  background-color: #fafafa;
}

<table
  class="c0"
  width="100%"
>
  <thead>
    <tr
      class="c1"
    >
      <th
        class="c2"
        colspan="8"
      >
        Time
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      class="c1"
    >
      <td
        class="c3"
      />
      <td
        class="c3"
      >
        Sun
      </td>
      <td
        class="c3"
      >
        Mon
      </td>
      <td
        class="c3"
      >
        Tue
      </td>
      <td
        class="c3"
      >
        Wed
      </td>
      <td
        class="c3"
      >
        Thu
      </td>
      <td
        class="c3"
      >
        Fri
      </td>
      <td
        class="c3"
      >
        Sat
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        Morning
        <br />
        10:00-13:00
      </td>
      <td
        class="c4"
        rowspan="3"
      >
        Day off
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          1
        </div>
        <div>
          Slave 
          1
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          2
        </div>
        <div>
          Slave 
          2
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          3
        </div>
        <div>
          Slave 
          3
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          4
        </div>
        <div>
          Slave 
          4
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          5
        </div>
        <div>
          Slave 
          5
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          6
        </div>
        <div>
          Slave 
          6
        </div>
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        Afternoon
        <br />
        14:00-18:00
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          1
        </div>
        <div>
          Slave 
          1
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          2
        </div>
        <div>
          Slave 
          2
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          3
        </div>
        <div>
          Slave 
          3
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          4
        </div>
        <div>
          Slave 
          4
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          5
        </div>
        <div>
          Slave 
          5
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          6
        </div>
        <div>
          Slave 
          6
        </div>
      </td>
    </tr>
    <tr
      class="c1"
    >
      <td
        class="c3"
      >
        Night
        <br />
        19:00-22:00
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          1
        </div>
        <div>
          Slave 
          1
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          2
        </div>
        <div>
          Slave 
          2
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          3
        </div>
        <div>
          Slave 
          3
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          4
        </div>
        <div>
          Slave 
          4
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          5
        </div>
        <div>
          Slave 
          5
        </div>
      </td>
      <td
        class="c3"
      >
        <div>
          Master 
          6
        </div>
        <div>
          Slave 
          6
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
